Feature/notifications util

mateuszRybczonek opened the pull request:
This PR adds:

Linters:
- eslint config and rules
- prettier for code formatting

Reusable components:
- TheOverlay - a single overlay component that lives at the App.vue level and can be triggered from any other place in the app
- BaseModal - a reusable wrapper component that contains common styles for the modal and uses props <--> emits instead of passing a callback method

It has no breaking changes, therefore can be safely merged into `develop`

mateuszRybczonek committed:
feat(overlay): implement single overlay logic

mateuszRybczonek committed:
feat(modal): extract common modal logic to BaseModal.vue

mateuszRybczonek committed:
feat(modal): set default title value

mateuszRybczonek committed:
chore: revert vue upgrade

mateuszRybczonek committed:
chore: fix unit tests

mateuszRybczonek committed:
chore: commit yarn.lock

mateuszRybczonek committed:
chore: add eslint and prettier

mateuszRybczonek committed:
chore: fix indentations

JettTech reviewed (approved):
LGTM.  Nice work.  One request I have is that we have at least one dev from holodev americas also give a thumbs up as this is updating shared code largely based on the americas' codebase.  (Maybe we could also bring up the topic of reviewers at our next ui guild.)

Paterick reviewed (approved):
That looks great Mateusz! Thanks for the clean-up and the standardized modal.

I like all the style changes except putting the attributes and properties on separate lines. In my opinion it makes the html unwieldy, causes lots of scrolling, and makes it harder to determine indentation. I think there are times when I have a bunch of attributes it helps readability but I don't think we need a hard rule to always do it.  I know everyone has different opinions about this but I vote against it.

robbiecarlton commented on src/utils/notifications.js:
```diff
@@ -0,0 +1,34 @@
+  hideBanner: 'hideBanner'
+}
+
+export const EOverlayType = {
+  loading: 0,
+  message: 1
+}
```
without typescript, I'm not sure what's the value in this kind of constant. There's nothing in js that prevents a typo (`if this.overlaytype === EOverlayType.messsage`), so why not just use the actual string 'loading', 'message' etc.

mateuszRybczonek replied:
It actually prevents typos, it depends what IDE do you use but it works with autocompletion (I use Webstorm). One benefit is typos prevention, the other is the autocompletion so I don't have to remember what options are available. It really speeds up the development in the longer term in my opinion.

<img width="769" alt="Screenshot 2022-06-23 at 11 36 11" src="https://user-images.githubusercontent.com/17565389/175268107-b3a3bc11-6363-4def-b9fb-3a3ba21cc8c0.png">

About Typescript, I agree, I could use strings here instead of numbers, but when we switch to Typescript at some point (and I really think we should someday) it will be much easier to change that const into an enum.

```
export enum EOverlayType {
  loading,
  message
}
```

robbiecarlton commented on src/pages/HappDetails.vue:
```diff
@@ -1,23 +1,47 @@
-          <div class='description'>{{ happ.description }}Happ to make you more productive.</div>
+      <div class="columns">
+        <div class="left-column desktop">
+          <HappImage
+            :happ="happ"
+            size="178px"
+            class="happ-image"
+          />
```
I think this is less readable than it was before.

robbiecarlton commented on src/pages/HappDetails.vue:
```diff
@@ -34,54 +58,73 @@
     </div>
-    <StopHostingModal v-if="hostingModalVisible" :handleClose="closeHostingModal" :stopHostingHapp="stopHostingHapp" :happName="happ.name" />
+
+    <StopHostingModal
+      :is-visible="hostingModalVisible"
+      :happ-name="happ.name"
+      @close="closeHostingModal"
+      @stop-hosting-happ="stopHostingHapp"
+    />
```
but this is more readable.

robbiecarlton commented on .eslintrc.js:
```diff
@@ -0,0 +1,334 @@
+          'error',
+          {
+            singleline: {
+              max: 1
```
```suggestion
              max: 3
```

robbiecarlton commented on src/App.vue:
```diff
@@ -1,26 +1,50 @@
+const theOverlay = ref()
+
+function showBusyOverlay(object) {
+  theOverlay.value?.show(object)
```
This is cool, didn't know you could do this

robbiecarlton commented on .eslintrc.js:
```diff
@@ -0,0 +1,334 @@
+    blankLine: 'always',
+    prev: '*',
+    next: 'for'
+  },
+  {
+    blankLine: 'always',
+    prev: '*',
+    next: 'function'
+  },
+  {
+    blankLine: 'always',
+    prev: '*',
+    next: 'if'
+  },
+  {
+    blankLine: 'always',
+    prev: '*',
+    next: 'iife'
+  },
+  {
+    blankLine: 'always',
+    prev: '*',
+    next: 'switch'
+  },
+  {
+    blankLine: 'always',
+    prev: 'case',
+    next: 'case'
+  },
+  {
+    blankLine: 'always',
+    prev: '*',
+    next: 'try'
+  },
+  {
+    blankLine: 'always',
+    prev: '*',
+    next: 'while'
+  }
+]
```
I would rather just get rid of this. It's not solving a problem I can see. I'd rather err on the side of fewer linting rules

robbiecarlton commented on .eslintrc.js:
```diff
@@ -0,0 +1,334 @@
+  }
+]
+const kBaseRules = {
+  'accessor-pairs': 'error',
```
same as above

robbiecarlton reviewed (changes requested):
Overlay looks good, a bunch of the clean up looks good.

The event emitter pattern for the overlay is cool, I haven't done that before. I would just use a store to do it, but this way has less boiler plate

I agree with Patrick about breaking attributes onto multiple lines. I think doing that specifically when it improves readability is great, but as a blanket practice I think it hurts readablity.

Also, there were a couple of linting rules that I commented on but then I stopped because I think there's a broader conversation to be had. In the past we've used standard js and I'm not sure why we need stricter rules than that. Let's discuss this at the next design guild.

I'm happy for you to merge this without the new linting rules.

mateuszRybczonek committed:
chore: reduce linter rules

mateuszRybczonek commented:
I agree, maybe I went a bit too extreme with the rules but I really think we shouldn't rely only on standard rules plus there were no rules for Vue in the past.

@robbiecarlton I removed the less critical rules from the config and left only those that I really think are necessary. I also split the into error-prevention and style-consistency rules so we have a bit more understanding about them. Please have a look again.

To support those rules see below some things that were caught by the new rules.

1. CSS props reassign:
<img width="256" alt="CSS props reassign" src="https://user-images.githubusercontent.com/17565389/175281872-40156720-eae9-4d25-b257-c53953d6d79a.png">

2. Magic numbers without meaningful names
OK, this one I can figure out
<img width="590" alt="Magic numbers 1" src="https://user-images.githubusercontent.com/17565389/175281879-5ba791f7-e059-4d42-9d65-2c2170e8a7d1.png">
But this I have no idea why it is 4, for me looks like a random number, why not 5 or 6? It should really be explained why is it 4.
<img width="791" alt="Magic numbers 2" src="https://user-images.githubusercontent.com/17565389/175281883-dabc2445-97db-4d81-befe-c89e2cbe0d57.png">

Our `<Button>` vs. html `<button>`
https://vuejs.org/style-guide/rules-essential.html#use-multi-word-component-names
<img width="767" alt="Multi word component name" src="https://user-images.githubusercontent.com/17565389/175281887-9549f7d7-cf39-4fd9-beb8-5d83b8e7f666.png">

3. Props should always be defined as `required` or have a `default` value just for us to know what is required.
https://vuejs.org/style-guide/rules-essential.html#use-detailed-prop-definitions
<img width="701" alt="Props definition" src="https://user-images.githubusercontent.com/17565389/175281891-ec0b2ea7-0891-4dbe-883e-65d1387d39a0.png">

4. Hmm... self-explanatory
<img width="561" alt="Spacings" src="https://user-images.githubusercontent.com/17565389/175281894-a565ced0-1e75-42e3-892a-37620b2050a1.png">

5. Template readability (scrolling vertically vs. scrolling vertically and horizontally)
<img width="1257" alt="Template readability" src="https://user-images.githubusercontent.com/17565389/175281896-c6e53bc3-ac3f-4da6-8884-3f1e1cb5a75f.png">

6. Unused imports
<img width="760" alt="Unused imports" src="https://user-images.githubusercontent.com/17565389/175281902-a01f86dc-ceff-48e4-ab82-026fb3cfe4f9.png">

7. Unused CSS classes
<img width="562" alt="Unused selectors" src="https://user-images.githubusercontent.com/17565389/175281904-425b0432-159d-4f70-80fe-b5f14f46603b.png">

I think we can get rid of all those issues very easily and prevent them from occurring in the future by applying more strict ESLint/Prettier rules that I propose. We can still discuss the style rules (I have nothing against changing the max lines from 1 to 3) but I think we should really put more pressure on the quality of out code.
